[inductor] Add a C shim layer for libtorch

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109436
* #109351
* __->__ #109022

Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Differential Revision: [D49149582](https://our.internmc.facebook.com/intern/diff/D49149582)

desertfire committed:
[inductor] Add a C shim layer for libtorch

Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Add a C shim layer for libtorch"


Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49149582).

desertfire committed:
Update on "[inductor] Add a C shim layer for libtorch"


Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49149582](https://our.internmc.facebook.com/intern/diff/D49149582)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49149582).

desertfire committed:
Update on "[inductor] Add a C shim layer for libtorch"


Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49149582](https://our.internmc.facebook.com/intern/diff/D49149582)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49149582).

desertfire committed:
Update on "[inductor] Add a C shim layer for libtorch"


Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49149582](https://our.internmc.facebook.com/intern/diff/D49149582)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49149582).

desertfire committed:
Update on "[inductor] Add a C shim layer for libtorch"


Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49149582](https://our.internmc.facebook.com/intern/diff/D49149582)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49149582).

desertfire committed:
Update on "[inductor] Add a C shim layer for libtorch"


Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49149582](https://our.internmc.facebook.com/intern/diff/D49149582)

[ghstack-poisoned]

ezyang commented on torch/csrc/inductor/aoti_torch/shim_cuda.cpp:
```diff
@@ -0,0 +1,28 @@
+  try {                                                      \
+    __VA_ARGS__                                              \
+  } catch (const std::exception& e) {                        \
+    std::cerr << "Error: " << e.what() << std::endl;         \
```
Wondering if this should be sent to some real logging framework 🤔  but what about ABI .......

ezyang replied:
oh but this is in the predictor, you have access to everything as needed

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,177 @@
+using AtenTensorHandle = AtenTensorOpaque*;
+
+struct AOTITensorManagerOpaque {};
+using AOTITensorManagerHandle = AOTITensorManagerOpaque*;
```
docs for these structs would be appreciated. Also, I wouldn't make the struct have an empty body; that implies it's a zero-size struct. Instead, you just want to have a forward declaration only (but no body) so that dereferencing the pointer is illegal.

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,177 @@
+extern "C" {
+#endif
+
+// WARNING: Change the following signatures will break ABI compatibility
```
docs for all the functions too

ezyang replied:
All of the functions need to document their ownership semantics (do they return borrowed or new references?)

desertfire replied:
Added comments before functions that allocate new tensors

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,177 @@
+using AOTITensorManagerHandle = AOTITensorManagerOpaque*;
+
+enum AOTITorchError : int32_t {
+  Success = 0,
```
This dumps an identifier named Success in your namespace (because this is not an enum class). You need to manually namespace it

desertfire replied:
changed to macros

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,177 @@
+AOTI_TORCH_EXPORT AOTI_TORCH_NOINLINE int32_t aoti_torch_dtype_int64();
+
+AOTI_TORCH_EXPORT AOTI_TORCH_NOINLINE AOTITensorHandle
+aoti_torch_optional_tensor();
```
...what's an optional tensor 🤔

ezyang replied:
Apparently this is generating a nullopt optional tensor, which is pretty weird, but @desertfire tells me this is a preexisting condition (still very weird)

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,177 @@
+AOTI_TORCH_EXPORT AOTI_TORCH_NOINLINE AOTITorchError
+aoti_torch_create_tensor_manager(
+    AOTITensorManagerHandle* tensor_manager,
+    int64_t reserve_size);
```
I'm not really sure what a tensor manager is

desertfire replied:
Removed in the new version

desertfire commented:
The generated code for `test_aot_inductor.test_simple` looks like the following,

//Before
```
void AOTInductorModel::run_impl(
    const std::vector<at::Tensor>& args,
    std::vector<at::Tensor>& outputs,
    cudaStream_t stream,
    ProxyExecutor* proxy_executor) {
    auto arg1_1 = args[0];
    auto arg2_1 = args[1];
    auto L__self___weight = constants_->at("L__self___weight");

    c10::optional<at::Scalar> optional_scalar;
    c10::optional<c10::string_view> optional_string;
    c10::optional<at::Layout> optional_layout;
    c10::optional<at::Tensor> optional_tensor;
    torch::List<c10::optional<at::Scalar>> optional_list;

    at::cuda::CUDAStreamGuard stream_guard(at::cuda::getStreamFromExternal(stream, 0));
    auto buf0 = outputs[0];
    // Source Nodes: [add], Original ATen: [aten.add]
    at::addmm_out(buf0, arg1_1, arg2_1, reinterpret_tensor(L__self___weight, {10L, 10L}, {1L, 10L}, 0L), 1, 1);
    arg1_1.reset();
    arg2_1.reset();
}
```

// After
```
void AOTInductorModel::run_impl(
        std::vector<at::Tensor>& args,
        std::vector<at::Tensor>& outputs,
        cudaStream_t stream,
        ProxyExecutor* proxy_executor) {
    AOTITensorManagerHandle aoti_tensor_manager;
    aoti_torch_create_tensor_manager(&aoti_tensor_manager, 4);
    AOTITensorHandle arg1_1;
    aoti_torch_extern_tensor_to_aoti_tensor(aoti_tensor_manager, &arg1_1, reinterpret_cast<AtenTensorHandle>(&args[0]));
    AOTITensorHandle arg2_1;
    aoti_torch_extern_tensor_to_aoti_tensor(aoti_tensor_manager, &arg2_1, reinterpret_cast<AtenTensorHandle>(&args[1]));
    AOTITensorHandle L__self___weight;
    aoti_torch_extern_tensor_to_aoti_tensor(aoti_tensor_manager, &L__self___weight, reinterpret_cast<AtenTensorHandle>(&constants_->at("L__self___weight")));

    aoti_torch_set_current_cuda_stream(stream, 0);
    AOTITensorHandle output_aoti_tensor_0;
    aoti_torch_extern_tensor_to_aoti_tensor(aoti_tensor_manager, &output_aoti_tensor_0, reinterpret_cast<AtenTensorHandle>(&outputs[0]));
    auto buf0 = output_aoti_tensor_0;
    // Source Nodes: [add], Original ATen: [aten.add]
    AOTITensorHandle aoti_tensor_0;
    int64_t int_array_0[] = {10L, 10L};
    int64_t int_array_1[] = {1L, 10L};
    aoti_torch__reinterpret_tensor(aoti_tensor_manager, &aoti_tensor_0, L__self___weight, 2, int_array_0, int_array_1, 0L);
    aoti_torch_addmm_out(buf0, arg1_1, arg2_1, aoti_tensor_0, 1L, 1L);

    aoti_torch_destroy_tensor_manager(aoti_tensor_manager);
}
```

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,177 @@
+#endif
+
+struct AOTITensorOpaque {};
+using AOTITensorHandle = AOTITensorOpaque*;
```
this is at::Tensor* under the hood

ezyang commented:
Call graph: Predictor ~> model.so ~> Predictor shim

ezyang commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -0,0 +1,528 @@
+// TODO: implement a more efficient version instead of calling into aten
+AOTITorchError aoti_torch_empty_strided(
+    AOTITensorManagerHandle tensor_manager,
+    AOTITensorHandle* out,
```
It's weird that the output argument is not at the front or end (which would be conventional in C)

ezyang commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -0,0 +1,528 @@
+    TORCH_CHECK(
+        success,
+        "Unexpected deleter function on storage, could not swap function");
+    // (void)success; // get around the unused variable warning
```
This stuff is going to go away in the new agreement

desertfire committed:
Update on "[inductor] Add a C shim layer for libtorch"


Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49149582](https://our.internmc.facebook.com/intern/diff/D49149582)

[ghstack-poisoned]

desertfire commented on torch/_inductor/codegen/aot_runtime/interface.cpp:
```diff
@@ -57,7 +57,7 @@ AOTInductorError AOTInductorModelContainerRun(
           container_handle);
 
-  const auto* inputs = reinterpret_cast<const at::Tensor*>(inputs_handle);
+  auto* inputs = reinterpret_cast<at::Tensor*>(inputs_handle);
```
Yes. Doing const_cast back and forth makes it meaningless to pass as const in the first place.

desertfire commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,177 @@
+using AOTITensorHandle = AOTITensorOpaque*;
+
+struct AtenTensorOpaque {};
+using AtenTensorHandle = AtenTensorOpaque*;
```
Having AtenTensorOpaque makes it easy to do `using RAIIAtenTensorHandle = std::shared_ptr<AtenTensorOpaque>;` in model.h

desertfire commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -0,0 +1,528 @@
+// This is NOT meant to be a general replacement of at::Tensor. The design
+// choices made here are highly coupled with AOTInductor's codegen, especially
+// with respect to resource management.
+// 1. Because AOTInductor codegen does static memory planning, we don't need to
```
Not relevant anymore in the new implementation

desertfire commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -0,0 +1,528 @@
+};
+
+// A placeholder for optional tensor
+AOTITensor optional_tensor_(
```
The current codegen emits something like `c10::optional<at::Tensor> optional_tensor;` to satisfy functions that missing passing argument for optional parameter. I left this part out in the new code since it is not needed for our initial models. Will revisit.

desertfire commented:
New generated code for test_aot_inductor.test_buffer_reuse:

```
// Before
void AOTInductorModel::run_impl(
    const std::vector<at::Tensor>& args,
    std::vector<at::Tensor>& outputs,
    cudaStream_t stream,
    ProxyExecutor* proxy_executor) {
    auto arg0_1 = args[0];
    auto arg1_1 = args[1];

    at::cuda::CUDAStreamGuard stream_guard(at::cuda::getStreamFromExternal(stream, 0));
    auto buf0 = at::empty_strided({4L, 4L}, {4L, 1L}, at::TensorOptions(c10::Device(at::kCUDA, 0)).dtype(at::kFloat));
    // Source Nodes: [a], Original ATen: [aten.sin]
    if (triton_poi_fused_sin_0 == nullptr) {
          triton_poi_fused_sin_0 = loadKernel("/tmp/torchinductor_binbao/clzryizwlh23a7eynokc2m7vu5y45sg4c26rotsp437ictdjgvdm/cq3l4dhwisbz5a4z54d2nmdjcqapnh4stdihv6cnclxkt2gyptp3.cubin", "triton_poi_fused_sin_0_0d1d2d", 0);
    }
    CUdeviceptr var_0 = reinterpret_cast<CUdeviceptr>(arg0_1.data_ptr());
    CUdeviceptr var_1 = reinterpret_cast<CUdeviceptr>(buf0.data_ptr());
    auto var_2 = 16;
    void* kernel_args_var_0[] = {&var_0, &var_1, &var_2};
    Grid triton_poi_fused_sin_0_grid_0 = Grid(1L, 1L, 1L);
    launchKernel(triton_poi_fused_sin_0, triton_poi_fused_sin_0_grid_0.grid_x, triton_poi_fused_sin_0_grid_0.grid_y, triton_poi_fused_sin_0_grid_0.grid_z, 1, 0, kernel_args_var_0, stream);
    auto buf1 = at::empty_strided({4L, 4L}, {4L, 1L}, at::TensorOptions(c10::Device(at::kCUDA, 0)).dtype(at::kFloat));
    // Source Nodes: [b], Original ATen: [aten.cos]
    if (triton_poi_fused_cos_1 == nullptr) {
          triton_poi_fused_cos_1 = loadKernel("/tmp/torchinductor_binbao/clzryizwlh23a7eynokc2m7vu5y45sg4c26rotsp437ictdjgvdm/c57zfvg3rtvgpjin7hknj7jjyd46tph5n7iezq4ex4ndasj2esn4.cubin", "triton_poi_fused_cos_1_0d1d2d", 0);
    }
    CUdeviceptr var_3 = reinterpret_cast<CUdeviceptr>(arg1_1.data_ptr());
    CUdeviceptr var_4 = reinterpret_cast<CUdeviceptr>(buf1.data_ptr());
    auto var_5 = 16;
    void* kernel_args_var_1[] = {&var_3, &var_4, &var_5};
    Grid triton_poi_fused_cos_1_grid_1 = Grid(1L, 1L, 1L);
    launchKernel(triton_poi_fused_cos_1, triton_poi_fused_cos_1_grid_1.grid_x, triton_poi_fused_cos_1_grid_1.grid_y, triton_poi_fused_cos_1_grid_1.grid_z, 1, 0, kernel_args_var_1, stream);
    auto buf2 = at::empty_strided({4L, 4L}, {4L, 1L}, at::TensorOptions(c10::Device(at::kCUDA, 0)).dtype(at::kFloat));
    // Source Nodes: [a, b, c], Original ATen: [aten.cos, aten.mm, aten.sin]
    at::mm_out(buf2, buf0, buf1);
    buf0.reset();
    buf1.reset();
    auto buf3 = buf2; buf2.reset();  // reuse
    // Source Nodes: [d, e], Original ATen: [aten.relu, aten.sigmoid]
    if (triton_poi_fused_relu_sigmoid_2 == nullptr) {
          triton_poi_fused_relu_sigmoid_2 = loadKernel("/tmp/torchinductor_binbao/clzryizwlh23a7eynokc2m7vu5y45sg4c26rotsp437ictdjgvdm/cqwdyklivtkfvpt5xa5fju2lojcnsuona6jze2wdkocjiq5lawen.cubin", "triton_poi_fused_relu_sigmoid_2_0d1d", 0);
    }
    CUdeviceptr var_6 = reinterpret_cast<CUdeviceptr>(buf3.data_ptr());
    auto var_7 = 16;
    void* kernel_args_var_2[] = {&var_6, &var_7};
    Grid triton_poi_fused_relu_sigmoid_2_grid_2 = Grid(1L, 1L, 1L);
    launchKernel(triton_poi_fused_relu_sigmoid_2, triton_poi_fused_relu_sigmoid_2_grid_2.grid_x, triton_poi_fused_relu_sigmoid_2_grid_2.grid_y, triton_poi_fused_relu_sigmoid_2_grid_2.grid_z, 1, 0, kernel_args_var_2, stream);
    auto buf4 = outputs[0];
    // Source Nodes: [d, e, g], Original ATen: [aten.add, aten.relu, aten.sigmoid]
    at::addmm_out(buf4, buf3, arg0_1, arg1_1, 1, 1);
    arg0_1.reset();
    arg1_1.reset();
}
```
and
```
// After
void AOTInductorModel::run_impl(
    std::vector<at::Tensor>& args,
    std::vector<at::Tensor>& outputs,
    cudaStream_t stream,
    ProxyExecutor* proxy_executor) {
    auto arg0_1 = create_raii_tensor_handle_for_extern(reinterpret_cast<AtenTensorHandle>(&args[0]));
    auto arg1_1 = create_raii_tensor_handle_for_extern(reinterpret_cast<AtenTensorHandle>(&args[1]));
    auto output_tensor_handle_0 = create_raii_tensor_handle_for_extern(reinterpret_cast<AtenTensorHandle>(&outputs[0]));

    AOTI_TORCH_ERROR_CHECK(aoti_torch_set_current_cuda_stream(stream, 0));
    int64_t int_array_0[] = {4L, 4L};
    int64_t int_array_1[] = {4L, 1L};
    AtenTensorHandle buf0_handle;
    AOTI_TORCH_ERROR_CHECK(aoti_torch_empty_strided(&buf0_handle, 2, int_array_0, int_array_1, aoti_torch_dtype_float32(), aoti_torch_device_type_cuda(), 0));
    auto buf0 = create_raii_tensor_handle_for_temp(buf0_handle);
    // Source Nodes: [a], Original ATen: [aten.sin]
    if (triton_poi_fused_sin_0 == nullptr) {
         triton_poi_fused_sin_0 = loadKernel("/tmp/torchinductor_binbao/clzryizwlh23a7eynokc2m7vu5y45sg4c26rotsp437ictdjgvdm/cq3l4dhwisbz5a4z54d2nmdjcqapnh4stdihv6cnclxkt2gyptp3.cubin", "triton_poi_fused_sin_0_0d1d2d", 0);
    }
    CUdeviceptr var_0;
    AOTI_TORCH_ERROR_CHECK(aoti_torch_get_data_ptr(arg0_1.get(), reinterpret_cast<void**>(&var_0)));
    CUdeviceptr var_1;
    AOTI_TORCH_ERROR_CHECK(aoti_torch_get_data_ptr(buf0.get(), reinterpret_cast<void**>(&var_1)));
    auto var_2 = 16;
    void* kernel_args_var_0[] = {&var_0, &var_1, &var_2};
    launchKernel(triton_poi_fused_sin_0, 1, 1, 1, 1, 0, kernel_args_var_0, stream);
    AtenTensorHandle buf1_handle;
    AOTI_TORCH_ERROR_CHECK(aoti_torch_empty_strided(&buf1_handle, 2, int_array_0, int_array_1, aoti_torch_dtype_float32(), aoti_torch_device_type_cuda(), 0));
    auto buf1 = create_raii_tensor_handle_for_temp(buf1_handle);
    // Source Nodes: [b], Original ATen: [aten.cos]
    if (triton_poi_fused_cos_1 == nullptr) {
         triton_poi_fused_cos_1 = loadKernel("/tmp/torchinductor_binbao/clzryizwlh23a7eynokc2m7vu5y45sg4c26rotsp437ictdjgvdm/c57zfvg3rtvgpjin7hknj7jjyd46tph5n7iezq4ex4ndasj2esn4.cubin", "triton_poi_fused_cos_1_0d1d2d", 0);
    }
    CUdeviceptr var_3;
    AOTI_TORCH_ERROR_CHECK(aoti_torch_get_data_ptr(arg1_1.get(), reinterpret_cast<void**>(&var_3)));
    CUdeviceptr var_4;
    AOTI_TORCH_ERROR_CHECK(aoti_torch_get_data_ptr(buf1.get(), reinterpret_cast<void**>(&var_4)));
    auto var_5 = 16;
    void* kernel_args_var_1[] = {&var_3, &var_4, &var_5};
    launchKernel(triton_poi_fused_cos_1, 1, 1, 1, 1, 0, kernel_args_var_1, stream);
    AtenTensorHandle buf2_handle;
    AOTI_TORCH_ERROR_CHECK(aoti_torch_empty_strided(&buf2_handle, 2, int_array_0, int_array_1, aoti_torch_dtype_float32(), aoti_torch_device_type_cuda(), 0));
    auto buf2 = create_raii_tensor_handle_for_temp(buf2_handle);
    // Source Nodes: [a, b, c], Original ATen: [aten.cos, aten.mm, aten.sin]
    AOTI_TORCH_ERROR_CHECK(aoti_torch_mm_out(buf2.get(), buf0.get(), buf1.get()));
    buf0.reset();
    buf1.reset();
    auto buf3 = buf2; buf2.reset();  // reuse
    // Source Nodes: [d, e], Original ATen: [aten.relu, aten.sigmoid]
    if (triton_poi_fused_relu_sigmoid_2 == nullptr) {
         triton_poi_fused_relu_sigmoid_2 = loadKernel("/tmp/torchinductor_binbao/clzryizwlh23a7eynokc2m7vu5y45sg4c26rotsp437ictdjgvdm/cqwdyklivtkfvpt5xa5fju2lojcnsuona6jze2wdkocjiq5lawen.cubin", "triton_poi_fused_relu_sigmoid_2_0d1d", 0);
    }
    CUdeviceptr var_6;
    AOTI_TORCH_ERROR_CHECK(aoti_torch_get_data_ptr(buf3.get(), reinterpret_cast<void**>(&var_6)));
    auto var_7 = 16;
    void* kernel_args_var_2[] = {&var_6, &var_7};
    launchKernel(triton_poi_fused_relu_sigmoid_2, 1, 1, 1, 1, 0, kernel_args_var_2, stream);
    auto buf4 = output_tensor_handle_0;
    // Source Nodes: [d, e, g], Original ATen: [aten.add, aten.relu, aten.sigmoid]
    AOTI_TORCH_ERROR_CHECK(aoti_torch_addmm_out(buf4.get(), buf3.get(), arg0_1.get(), arg1_1.get(), 1L, 1L));
    arg0_1.reset();
    arg1_1.reset();
}
```

desertfire commented on torch/_inductor/config.py:
```diff
@@ -430,6 +430,9 @@ class aot_inductor:
     output_path = ""
 
+    # Wether to codegen abi compatible model.so
+    abi_compatible = True  # is_fbcode()
```
hard coded to True, I am working on adding test copy to test both and will change this back

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,157 @@
+
+#ifdef USE_CUDA
+// Assuming CUDA ABI is stable, and any major CUDA upgrade will require a
+// rebuild of model.so
```
I don't understand this comment?

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,157 @@
+// generated code in model.so, is responsible for wrapping those raw pointers
+// into RAIIAtenTensorHandle (see aot_runtime/model.h) which will take care of
+// freeing those allocated at::Tensor objects.
+struct AtenTensorOpaque {};
```
So what happened here. Does it not work to delete the ` {}` here?

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,157 @@
+using AtenTensorHandle = AtenTensorOpaque*;
+
+#ifdef __cplusplus
+extern "C" {
```
Why isn't the `extern "C"` include the struct definition above

desertfire replied:
fixed

ezyang committed:
comment updates on shim.h on "[inductor] Add a C shim layer for libtorch"


Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49149582](https://our.internmc.facebook.com/intern/diff/D49149582)

[ghstack-poisoned]

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,170 @@
+aoti_torch_tensor_copy_(AtenTensorHandle src, AtenTensorHandle dst);
+
+AOTI_TORCH_EXPORT AOTI_TORCH_NOINLINE AOTITorchError aoti_torch_addmm_out(
+    AtenTensorHandle out,
```
The naming convention here is very confusing, because we are using out to refer to two distinct concepts: (1) out pointer to be overwritten to "return" something from the function, and (2) non-pointer direct handle where we are going to write the data into the tensor. Need to disambiguate the names. `out` should be reserved for (2), so let's come up with some other name for (1). How about `ret`?

desertfire replied:
Good point. Fixed aoti_torch__reinterpret_tensor and aoti_torch_empty_strided.

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,170 @@
+// compatibility.  Typically, this means you will have to add a _v2 version
+// of a function that you, e.g., want to add a new function parameter to, and
+// maintain the old and new versions of the APIs until all old model.so
+// go out of use.
```
This comment is new, give it a read

ezyang commented on torch/csrc/inductor/aoti_torch/utils.h:
```diff
@@ -0,0 +1,15 @@
+  try {                                                      \
+    __VA_ARGS__                                              \
+  } catch (const std::exception& e) {                        \
+    std::cerr << "Error: " << e.what() << std::endl;         \
```
Put this in real logs with `LOG` macro instead of just dumping it to stderr

ezyang commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -0,0 +1,181 @@
+    const int64_t* strides_ptr,
+    int64_t offset_increment) {
+  AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
+    at::Tensor* self_tensor = reinterpret_cast<at::Tensor*>(self);
```
Consider making a helper function for this reinterpret cast

ezyang commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -0,0 +1,181 @@
+    at::Tensor* out_tensor =
+        new at::Tensor(std::move(torch::inductor::_reinterpret_tensor(
+            *self_tensor, sizes, strides, offset_increment)));
+    *out = reinterpret_cast<AtenTensorHandle>(out_tensor);
```
and the other way out as well. This will help avoid mistakes where you accidentally reinterpret cast something of the wrong type

ezyang commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -0,0 +1,181 @@
+AOTITorchError aoti_torch_delete_tensor_object(AtenTensorHandle tensor) {
+  AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
+    at::Tensor* t = reinterpret_cast<at::Tensor*>(tensor);
+    delete t;
```
Ahh I love the smell of tiny heap allocation in the morning (@desertfire tells me that this isn't really visible on the internal bench, but this is probably one of the first things I'd try to dump in microoptimizing. cc @swolchok)

ezyang commented on torch/csrc/inductor/aot_runtime/model_container.h:
```diff
@@ -125,7 +125,7 @@ class AOTInductorModelContainer {
 
   void run(
-      const std::vector<at::Tensor>& inputs,
+      std::vector<at::Tensor>& inputs,
```
how come?

desertfire committed:
Update on "[inductor] Add a C shim layer for libtorch"


Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49149582](https://our.internmc.facebook.com/intern/diff/D49149582)

[ghstack-poisoned]

desertfire commented on test/inductor/test_aot_inductor.py:
```diff
@@ -104,7 +115,18 @@ def run_multiple(
+    example_inputs,
+):
+    expected = model(*example_inputs)
+    with config.patch("aot_inductor.abi_compatible", self.abi_compatible):
```
I haven't turned on abi_compatible for every tests, and some of the fixes will come later.

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,167 @@
+// Get a pointer to the underlying storage data
+AOTI_TORCH_EXPORT AOTI_TORCH_NOINLINE AOTITorchError aoti_torch_get_data_ptr(
+    AtenTensorHandle tensor,
+    void** out_data_ptr // returns borrowed reference
```
this doesn't look fixed

ezyang commented on torch/csrc/inductor/aot_runtime/model.h:
```diff
@@ -323,5 +324,33 @@ class AOTInductorModel : public AOTInductorModelBase<AOTInductorModel> {
+// tensor, this RAII class will achieve a similarly short live range for
+// AtenTensorHandle in the ABI compatible mode.
+
+using RAIIAtenTensorHandle = std::shared_ptr<AtenTensorOpaque>;
```
You can land like this, but this is also easy optimization opportunity (as the uses of this handle should all be unique right?)

ezyang commented on torch/csrc/inductor/aot_runtime/model.h:
```diff
@@ -323,5 +324,33 @@ class AOTInductorModel : public AOTInductorModelBase<AOTInductorModel> {
+    AOTI_TORCH_ERROR_CHECK(
+        aoti_torch_delete_tensor_object(static_cast<AtenTensorHandle>(ptr)));
+  });
+}
```
The more conventional naming here would be "borrow" vs "steal" (former/latter)

ezyang commented on torch/csrc/inductor/aot_runtime/model.h:
```diff
@@ -323,5 +324,33 @@ class AOTInductorModel : public AOTInductorModelBase<AOTInductorModel> {
+  return RAIIAtenTensorHandle(handle, [](AtenTensorHandle ptr) {
+    // Do nothing for extern tensor handles
+  });
+}
```
It makes sense that you wanted to use this for arguments, but in fact the API for calling AOTInductor should *steal* the tensors passed into it. This ensures that tensors can be promptly deallocated as soon as AOTInductor is done using them, if there are no other references to them. If AOTInductor borrows input arguments, then they cannot deallocate until the very end. BTW, eager does it this way, that's what the .clear() on args list is for.

ezyang replied:
Bin tells me that upstream requested a borrow-only interface. That seems a bit strange to me. @zdevito do you know if TorchScript supports operators that steal their inputs?

ezyang replied:
Even if one integration wants to only send borrows to AOTInductor, I still think the model.so API should steal in tensors. If you don't do this, then it is impossible to support stealing without a stable API breakage. But if you take steals on model.so, the top level can always bump the refcount by one before sending the tensor in.

ezyang commented on test/inductor/test_aot_inductor.py:
```diff
@@ -18,6 +19,16 @@
+except unittest.SkipTest:
+    if __name__ == "__main__":
+        sys.exit(0)
+    raise
```
meh

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -156,10 +156,15 @@ def codegen(self, code: IndentedBuffer, device_cm_stack: contextlib.ExitStack):
-                )
+                if config.aot_inductor.abi_compatible:
+                    code.writeline(
+                        f"AOTI_TORCH_ERROR_CHECK(aoti_torch_set_current_cuda_stream(stream, {self.device_idx}));"
```
This doesn't seem like an exact equivalence; in particular, the old code would reset the stream back to its original setting at the end of execution (or on early exit.) Another RAII wrapper here would make it easy to replicate the correct semantics!

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -628,7 +630,7 @@ def codegen_python_shape_tuple(self, shape: Tuple[Expr, ...]) -> str:
         return self.codegen_python_shape_tuple(shape)
 
-    def codegen_reinterpret_view(self, name, size, stride, offset) -> str:
+    def codegen_reinterpret_view(self, name, size, stride, offset, writer) -> str:
```
What's a writer

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -959,6 +961,9 @@ def __init__(self):
         self.supports_intermediate_hooks = False
         self.outputs_need_copy = set()
         self.kernel_callsite_id = count()
+        self.int_array_id = count()  # for int array local variable declarations
```
You actually don't need the temp arrays, because you're in C++ and an initializer list `{1, 2}` will cast into a `const int*` just fine

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -985,12 +990,15 @@ def write_header(self):
+            self.header.splice(
+                """
+                #include <torch/csrc/inductor/inductor_ops.h>
+                #define reinterpret_tensor torch::inductor::_reinterpret_tensor
```
lol

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1078,16 +1096,31 @@ def write_wrapper_decl(self):
+                        )
+                    else:
+                        self.prefix.writeline(
+                            f"""auto {constants_key} = constants_->at("{constants_key}");"""
```
So what's the plan for constants?

ezyang replied:
Bin: It's going to morally passed in as an argument, so same as other args

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1078,16 +1096,31 @@ def write_wrapper_decl(self):
+                    self.prefix.writeline(
+                        f"auto {output_key} = create_raii_tensor_handle_for_extern("
+                        + f"reinterpret_cast<AtenTensorHandle>(&outputs[{idx}]));"
+                    )
```
any chance we can make the OSS version symmetric with this? (can be done later)

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1221,16 +1255,22 @@ def generate_return(self, output_refs):
 
     def generate_end(self, result):
         if V.graph.aot_mode:
+            result.writeline("}")
```
comment what this is ending

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1286,6 +1326,31 @@ def g(args):
             """
         )
 
+    def generate_c_shim_extern_kernel_call(self, kernel, args):
```
It would be good to have some sort of assert making sure you don't try to call kernels that aren't actually in the shim

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1286,6 +1326,31 @@ def g(args):
+        # In the abi_compatible mode, we call fallback aten ops through a C shim layer
+        kernel = "aoti_torch_" + kernel.split("::")[-1]
+        for i, arg in enumerate(args):
+            if arg.startswith("create_raii_tensor_handle_for_temp("):
```
meh

ezyang replied:
You can avoid needing to do the explicit `.get()` call by adding an implicit conversion to your RAII class, `operator AtenTensorHandle()`

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1286,6 +1326,31 @@ def g(args):
+                self.codegen_int_array_var(size),
+                self.codegen_int_array_var(stride),
+            ] + args
+            self.generate_c_shim_extern_kernel_call(kernel, args)
```
this apparently needs some more work

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1347,18 +1416,39 @@ def generate_profiler_mark_wrapper_call(self, stack):
+        # of self.generate), the writeline behavior is different in the two passes.
+        # As a result, the emitted int array declarations may appear in a later
+        # position of the generated code, so the second pass codegen should not
+        # reuse int array declarations generated in the first pass
```
It would be helpful to say when writer is None versus not

ezyang reviewed (approved):
I don't really understand the two pass memory planning thing, maybe grab someone who knows how that algo works. Check comments.

desertfire commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -0,0 +1,167 @@
+// The general guidelines for the C API:
+//
+//  - No exceptions, return an explicit error code to be checked at call site
+//  - Only pointers (AtenTensorHandle counts), integers and floats in headers
```
No, the plan is to not allow it. If we feel those are stable enough later, we can do that in v2.

desertfire commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1047,27 +1065,43 @@ def write_wrapper_decl(self):
                         f"auto {constants_key} = args[{constants_idx}];"
                     )
 
+            if V.graph.aot_mode and config.aot_inductor.abi_compatible:
```
Yes, it is.

desertfire committed:
Update on "[inductor] Add a C shim layer for libtorch"


Summary: This PR adds a limited C shim layer for libtorch. The ultimate goal is to ban any direct reference to aten/c10 data structures or functions, to avoid ABI breakage by providing stable C interfaces.

To make the review and landing easier, we break the changes into several steps. In this PR, we add C interfaces for certain libtorch functions and modify the wrapper codegen to generate calls to those interfaces. There are a few major items to be addressed in future PRs:

* The AOTInductor runtime interface still takes lists of aten tensors as input and output
* The interaction with ProxyExecutor (general fallback support) needs to move away from aten tensor
* Remove all references to aten/c10 headers in the AOTInductor-generated code
* Bring up the passrate for OSS models as there are more libtorch functionalities to be supported

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49149582](https://our.internmc.facebook.com/intern/diff/D49149582)

[ghstack-poisoned]

desertfire commented:
merged in another exported PR, https://github.com/pytorch/pytorch/pull/109391
